### ISSUE TITLE
Code commenting

### ISSUE THREAD

**Issue by majikandy** [stratisproject/StratisBitcoinFullNode#1400]
There is a very strict policy on code comments on this project.

However, a lot of the motivation for this, whilst good, is quite outdated in its ideas.

In the policy :- 
https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/Documentation/commenting-policy.md

>Yes, it is true that you should always try to write the code as if you could not write any comment to it. 

The above is something I think we can all agree on.

However, the motivation section mentions many things as facts, which aren't actually facts.

Eg.

>Having a well defined code style helps with that. 

This is debatable - a strict style like the one on this project actually has many areas that force noise and clutter into the code - e.g. forcing some summaries that repeat what the code already says.

>But regardless how good your code is, you still need to add comments. 

This is patently NOT a fact. You can have your cake and eat it and I believe you should try to, since code maintenance is more important than the first time it is written. Code is read many many more times than it is written - therefore the extra effort that is used for commenting, could be used to make the code even more expressive.

There are a huge number of developers who believe that comments are a "last resort" after you have completely failed to communicate the intent in code alone.

The main resource on this would be Robert C Martin and mainly in his book Clean Code.

This is not to say we should never have comments - that would be a mistake, there are plenty of reasons to capture something in text that cannot be captured in code alone - however to enforce that we ALWAYS have to comment and write summaries just leads to unnecessary duplication, comments that are out of sync with the code, comments that mislead rather than help the next developer. 

However I think it is also fair to say that in some cases, they also help and save time. But I stand by that being a last resort and not a first course of action.

**Comment by noescape00** [comment]
I think this rule indeed needs to be lose for some particular case: 

When we have a class A and it has a summary comment we do not need to comment every instance of that class (unless dev want to for some reason). Currently we do so and it just leads to copying the summary comment over and over.

**Comment by Aprogiena** [comment]
I'm happy to relax the requirement for fields that are used in other components commonly for the same thing and are injected in constructor via DI. Otherwise I believe we need comments everywhere (XMLDOC, not inline). Inline comments should be written if the code itself is not clear, I agree with that, no need to write inline comments if the code is clear. 

XMLDOC - i.e. comments of properties, methods, interfaces, classes, parameters, returns, exceptions ... they add another layer of security - code and comment must match, otherwise we have a bug in code, comment, or both. 

Each class should also have a big comment explaining its role from bigger picture point of view. And again, this is another layer for bug prevention - if big picture does not match the implementation, it is a bug. 

Not only it is extra layer of protection, but it is also helpful for intellisense, which speeds up the coding.

**Comment by Aprogiena** [comment]
So that is the 1st part of the argument, the 2nd one is about providing information to developers that directly uses the code. I will add this later.

**Comment by dangershony** [comment]
I would like to remove the need to comment constructors and properties, properties are often just repeating comments, there is mostly not befit there and if a user needs documentation navigating to the property's class for it's comments is a click away.

I would also like to remove the need to comment on method parameters same argument applies to properties.

I find myself writing many times the comment like 
```
        /// <param name="network">The blockchain network.</param>
```

**Comment by Aprogiena** [comment]
I would be okay with relaxing the parameter's comments for such parameters that are used in many components in a same way.

**Comment by Aprogiena** [comment]
So as for the providing information to developers. It is very handy and sometimes crucial to provide information to a fellow developer who is using the code either from external component or from within the component we are writing.

As we write the code now, the comments are the best place to put this information to. Should we ever start using whatever tools for formal validation or static analysis of the code, we may find ourselves moving such information from the comments to some special tags or attributes.

This information includes but is not limited to: locking contracts, parameter restrictions, contracts between parameters, exceptions thrown from the methods, discussion about time and space complexity, links to resources related to the code, et cetera

It is impractical to try to put all this information inside the names, although when possible, the most critical information should be included – such as the requirement of holding locks before calling a certain method is very nice to have on the method name.

**Comment by dangershony** [comment]
I am a believer of writing code that is so readable that comments are not needed, however I am also a bit skeptic if the ability to achieve that.

I would love to see examples of code that is so readable that comments are not needed.
Will this be a different way of writing code? or just more focus on type/method names.
We see the new test specification code where method high level expresses a lot of intent in plain English I wonder if such similar approaches are the intent of readable code.

**Comment by majikandy** [comment]
>... they add another layer of security - code and comment must match, otherwise we have a bug in code, comment, or both.

Tests are a far better solution to this problem. Tests are the live documentation for the system that express how the system works. If there is no test, how did the code get written in the first place and accepted? putting that aside....  if you are changing a piece of code, it is your duty to write tests that prove it works. Ideally TDD, but let's save that one for another day.

So you say code and comment must match but you aren't allowed to have a redundant comment where the code is clear? - Those two statements appear to compete with each other.

Personally I am skeptical of all comments and will read the tests and the code to understand what it does - it sounds like you are also skeptical of the comment, otherwise why would you be comparing the comment and the code to see if there is a bug?  Do you trust the code or the comment? What do you do to resolve it? you look into it and you fix them to match - thus causing it to now be a redundant comment.

A quick google of "code comments add a layer of security to your code" returns no obviously relevant results.  If this was an industry accepted practice, wouldn't we have thousands of results?
 
>This information includes but is not limited to: locking contracts, parameter restrictions, contracts between parameters, exceptions thrown from the methods, discussion about time and space complexity, links to resources related to the code, et cetera

Some of those things, yes occasionally very useful - time and space complexity, sure, why not, especially if we have chosen a surprising one in that context.  Links to resources, of course include these.

Parameter restrictions, better solved with things like Value objects rather than using/abusing primitives like int or string.

>I would love to see examples of code that is so readable that comments are not needed.

@dangershony Here are some quick examples:-

Article by Jeff Atwood - https://blog.codinghorror.com/coding-without-comments/

The following is from Clean Code by Robert C. Martin
```
Which would you rather see? This:

// Check to see if the employee is eligible for full benefits 
if (( employee.flags & HOURLY_FLAG) && (employee.age > 65))

Or this?

if (employee.IsEligibleForFullBenefits())

It takes only a few seconds of thought to explain most of your intent in code.
```
- I would add to this that our coding policy rules say we need to add a summary to the private method too, which would completely ruin the elegance of having the private method. The summary would literally repeat the code inside and therefore break DRY at the very least.


>Will this be a different way of writing code? or just more focus on type/method names.

Yes focus on names, but more than that - 
1. Get obsessive about Single Responsibility Principle and other SOLID principles
2. Use methods at the same level of abstraction
3. Pronouncable names to aid with communication
4. Avoiding acronyms!!  Does sc mean side chain? or smart contract?
5. Not using similar variable names as another one near it just because you needed a second variable
6. Use names that reflect the system domain, the context, and the problems that must be solved
7. Use names with words that say what it really does
8. The optimal number of parameters of a method is zero, after one and two. Three should be avoided, but if you think it should be used, have a good justification - *This includes constructor dependency injection*
9. A parameter of a boolean, already says a method does more than one thing
10. Follow law of demeter and only talk to your friends, not your friends friends
11. Use exceptions only in exceptional circumstances
12. Don't guard for null - make sure it *NEVER* reaches that point in the code with a null. Never pass null to methods and never return null from methods, prefer empty objects instead.
13.  Use TDD to drive the code, doing the test first forces you to not write complicated code, partly because it is really hard to write complicated code when you are writing tests to cover it on the fly as you go.
    i. Don't create code before you have a failing test.
    ii. Don't create more tests than necessary to fail. 
    iii You cannot write more code than enough to pass the test that is failing.


....Anyway that is far too much food for thought off the top of my head.

**Comment by Aprogiena** [comment]
>>    ... they add another layer of security - code and comment must match, otherwise we have a bug in code, comment, or both.

>Tests are a far better solution to this problem. Tests are the live documentation for the system that express how the system works. If there is no test, how did the code get written in the first place and accepted? putting that aside.... if you are changing a piece of code, it is your duty to write tests that prove it works. Ideally TDD, but let's save that one for another day.

Non-argument - I can say the same - `comments are better solution to this problem, they are live documentation ... if you are changing code it is your duty ...` - see ? it works as well

**Comment by Aprogiena** [comment]
> So you say code and comment must match but you aren't allowed to have a redundant comment where the code is clear? - Those two statements appear to compete with each other.

you are mixing inline comments and XMLDOC

**Comment by Aprogiena** [comment]
>  Do you trust the code or the comment? What do you do to resolve it?

Do you trust the test or the code? Again, the whole argumentation works for comments as well.

**Comment by Aprogiena** [comment]
> Parameter restrictions, better solved with things like Value objects rather than using/abusing primitives like int or string.

No, this is not about abusing types. There are often restrictions - e.g. something CAN be null, or something CANT be null, or something CAN be null ONLY if something else is TRUE, etc.

**Comment by dangershony** [comment]
Ah very good two points from the article you provided.

- You should always write your code as if comments didn't exist
- To write good comments you have to be a good writer

I agree on both accounts.

**Comment by majikandy** [comment]
@Aprogiena fair enough you can use the same argument for comments instead of tests, but why bother. Tests are runnable, comments aren't.

>Non-argument - I can say the same - comments are better solution to this problem, they are live documentation ... if you are changing code it is your duty ... - see ? it works as well

It almost works. Except a comment isn't live, it is free to slip away into being out of date and they 100% do slip and take a great deal of code review diligence to keep in sync

I can't believe I have to make these points, but by the same argument I guess neither can you.

**Comment by majikandy** [comment]
>you are mixing inline comments and XMLDOC

I am, why aren't you?  Why is an XMLDOC comment allowed to be redundant?

**Comment by bokobza** [comment]
I'm not going to comment on this, but as far as I'm concerned, this is a time suck and it's unacceptable.
We decided on these rules months ago after many many discussions.
What happened with not fixing something that is not broken?
Clearly we all have different views on the matter. Some like comments, some don't.

We need to concentrate on writing code and delivering rather than talking about writing code and delivering.

**Comment by Aprogiena** [comment]
> this is a time suck and it's unacceptable.

I am afraid you are right. I would still very much like anyone to question anything or propose anything, however, as you say, we must not waste our time. Therefore, assuming as it is obvious that certain proposal does not have an over helming majority of support, we should avoid spending more time on it.

Therefore not at any more comments here.

@majikandy I am still happy to have a Skype call and discuss this matter, if you are interested, but only outside of working hours – i.e. during weekends – to avoid wasting work time.

**Comment by dangershony** [comment]
>We decided on these rules months ago after many many discussions.

When we decided on the commenting policies there where only 3 of us, and I was happy to give it a try.
I am re-questioning that decision now, there are some very solid arguments that are made about not using a strict approach to writing comments and I would like to explore that.

I also think that blindly deciding that we should always put comments is not right, there are places and times where comments are useless/time-wasting but our policies will still demand that we write "something" down.

And again I repeat I am NOT against commenting I am just against the strict policy we currently have.

Looking at example in the dotnet core ns we see comments are optional.   
https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/File.cs

I would also argue that some devs are better at writing comments then others, so if a reviewer feels a comment is needed they can (a reviewer should understand the code they are reviewing and so be able to) provide a solid comment if they think the code is not clear from how its written.

I'd also agree to keep a requirement that classes be heavily commented on the intention of the class, such documentation almost always makes sense.

**Comment by dangershony** [comment]
I am going to propose this changes to the policy.

1 - Private properties don't require comments if name is descriptive enough or its clear form the type.
1.1 - Or we can use apros suggestion to put a special comment tag like //inheritdoc maybe //inheritparent (I did try to see if there is such an existing tag) that means look at the description of the type.
2 - Methods parameters don't require comments (or similar to 1.1)
3 - Constructors don't require comments.
4 - Private methods don't require comments if they are written in a way that its clear what they do.

Require in this context does not mean no comments at all but if the code is readable enough then comments be left out.

**Comment by bokobza** [comment]
First of all, a disclaimer: I'm happy for **anyone** to question **anything** at **anytime**. 
I believe this is how we move forward, innovate and better ourselves and what we do.

Having said that, after months-long discussions we arrived at a set of rules that define how we write code here at Stratis. Referring to my previous point, I'm happy for this to be challenged within reason.

However, I think some of us are missing the point with regards to **documentation**, **the use of var** and other **CONVENTIONS**.

No one really cares whether it's better or not to write full documentation or only when it's not obvious. No one cares whether we use var or not. The compiled code is exactly the same, I hope we all agree on that.

What these rules do is they **define a set of conventions between developers** in order to streamline the process and remove all manner of questions the developer asks himself in his daily work: "Should I do it? Is this the right thing to do? What will the reviewers think? etc..."
 
When as a rule you decide to have comments everywhere, or no comments at all, developers don't have to think/judge/assess whether comments on a particular element are required.
It removes all uncertainties and prevent all discussions between developers, i.e "we put documentation on everything". Done.
Having comments on everything + using var when type can be inferred and explicit type when not is actually the simplest options that let the developers concentrate on what's important.

If we allow to just do whatever, we're going to have endless discussions about whether a documentation is required or not, whether a private method requires documentation or not, whether the use of var is allowed or not etc...

I think relaxing the rules and allowing anyone to do as they please is a false economy.
Whatever we decide, it needs to be so obvious that no discussion between developers is possible.

To close, and I'm going to be blunt, if a developer can't write English, they need to learn. I ask for help whenever I need to, and I hope everyone feel they can, too. 
"I can't write English" is not a good excuse.

**Comment by fenix2222** [comment]
My 5 satoshis here:
I like @dangershony suggestion for a policy change. I don't see a point of putting useless comments that increase code reading time and add 0 value, I would actually argue a lot of the times it is an unnecessary maintenance burden.

If we look at Microsoft code (we could consider it to be a benchmark) or code from other big enterprise companies, comments are important (ie class header comments, important business logic decisions) but NOT mandatory in some cases.  

Currently, the issue is that I have already seen fair few developers complain about this policy which means it is a problem and maybe something that needs to be addressed. Every time we get a new developer joining Stratis we will see more and more discussion on this topic. 

As the last point, I would like to point out that unit tests have far greater value than comments. It provides a live documentation of the code that can be executed and improve confidence in the code. Having good test coverage ensures code quality while having good comments doesn't mean the code is of high quality nor it gives me any confidence when adding a new change that this change is not going to break other parts of the code.

**Comment by dangershony** [comment]
I want to stress again I am not about removing the rules just making then less strict, I am also in violation of some of the policies myself as a result.

For example @bokobza how strict our commenting rules are is this line in commit #1351 
https://github.com/stratisproject/StratisBitcoinFullNode/pull/1351/files#diff-119e0a2d82ee52a5f2018a9b4655e222R61

It violates **ADD-VALUE** policy.

**Comment by dangershony** [comment]
Violations also in this class which was reviewed by @Aprogiena 

https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/Stratis.Bitcoin/Base/BestChainSelector.cs#L73
